### ISSUE TITLE
Installer for Windows

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#5810]
It should install RethinkDB, maybe add a start menu entry for documentation, and register it as a service (it should ask whether to enable the service).

**Comment by bchavez** [comment]
I'm not sure there's a need for one. For me, I prefer a zip file as the way it's distributed now. Or at least keep this zip file option available, or `installer.exe /unzip c:\here`. My opinion might not mean much in the grand scheme of things, but I'd hope to point out:

Problems with installers:
* Adds friction
* Programs would normally install to `C:\Program Files\`. IIRC, the one "micro$oft" way would be to put only **program files & read-only data** in `C:\Program Files\`.... so `rethinkdb_data` files would have to go someplace else. But where? Maybe some default location like `C:\ProgramData` that is only shown in **Windows Explorer** when "show hidden files & folders" is configured in some obscure *View > Folder Options* setting dialog box. :crying_cat_face: 
* Multiple installations become a problem too. What does one do with MSI database entry in `appwiz.cpl` when multiple instances are installed? Does it uninstall the last instance that was installed? Or is a dialog box going to pop up and ask for which instance to uninstall?
* How do we handle updates to already installed instances?
* And what about that proxy mode? And multiple instances of proxy mode things?
* And about a custom UI for all the config flags? Join to/from cluster? Bind all/IP address?

:8ball: :8ball: :8ball: 

The only advantage I can think of for a MSI like installer is:
* Some adminy/MSI deployment scenario thing. But honestly, could just as easily be done with a custom PowerShell setup script.

I would try to keep it simple, maybe something like this:
```
rethinkdb.exe --install --myflags 
rethinkdb.exe --uninstall --myflags 
```
The `--install` and `--uninstall` flags are all that's needed to setup a **Windows Service** in the SCM in the current directory. Update? Boom :boom:. Replace the `.exe`. Need to roll back? Boom :boom:. Replace the `.exe`. Uninstall? Boom :boom:. Delete the directory (and if you want to remove from SCM **without the `rethinkdb.exe`**, that's completely cool too `sc delete <service name>` from the command prompt). No lingering footprint on the system.

---------
In contrast, w.r.t MSI installers:
> Imagine if we tried to uninstall RethinkDB with an MSI entry (and we accidentally deleted an install folder), the MSI installer would go **bat- - - -** crazy complaining that you deleted directory that is needed to complete the uninstall (lolz), so then uninstall would fail and a pesky & broken **RethinkDB** MSI entry would be present inside `appwiz.cpl`. So, to back out of your mistake, you'd need to ***install*** RethinkDB via MSI first (to recreate the missing files that were deleted), then run the `appwizl.cpl` ***remove program*** to finally ***uninstall*** RethinkDB properly.

> Or, use an **nirsoft** tool to edit the MSI database directly to remove the pesky **RethinkDB** MSI entry.

-------

RethinkDB is simple. One `.zip/.exe` file and we get a cool database without any drama. This is really nice. Since we are not installing OS-level drivers or anything fancy, I would maybe hold out against a GUI/MSI like installer until one is ***really*** needed. I dunno about the documentation shortcut either. Presumably, the documentation is *where* they got the installer from. ¯\_(ツ)_/¯

Just my 2 cents. :money_with_wings: 

Brian :sunglasses:

**Comment by danielmewes** [comment]
Thanks for your take on this @bchavez .

With an installer, we would still have a way for running `rethinkdb.exe` manually. Either we could add it to the PATH, or just document where it gets installed so that you can add it to the PATH manually.

Good question about where to store the database file and configuration for a global instance. If it was bound to a user, we could put it into "Application Data" or similar. But it isn't really.
I wonder how other databases on Windows do it (for example MS SQL Server).

The main benefit of providing an installer is that it's likely more familiar for new users. We can also bundle it up with some useful "getting started" links to documentation that we can add to the start menu.

Another interesting option for distribution is https://chocolatey.org/ by the way. I haven't looked into the details of that yet.

**Comment by danielmewes** [comment]
This is a great idea I think btw:
```
rethinkdb.exe --install --myflags 
rethinkdb.exe --uninstall --myflags 
```

**Comment by bchavez** [comment]
IIRC, MSSQL server goes against the grain by installing database instances in 
```
C:\Program Files\Microsoft SQL Server\MSSQL.InstanceID
```
And then...
##### [AppCert: Quote from Section 10.1](https://msdn.microsoft.com/en-us/library/mt674655%28v=vs.85%29.aspx)
> For native 32-bit and 64-bit apps in %ProgramFiles%, and %ProgramFiles(x86)% for 32-bit apps running on x64. ***User data or app data must never be stored in this location because of the security permissions configured for this folder.***

¯\\_ (ツ)_/¯

Just hope we can keep a nice, simple, no frills, no drama, `.zip+.exe` :package: around.
Totally, `rethinkdb.exe --install ServiceName --andAllMyFlags` would be awesome. :star2:

**Comment by ghiscoding** [comment]
I am also not so much interested in the MSI installer but seriously can't wait to see the Service install. The idea of starting a Service instead of always having a shell open just for RethinkDB would be so much more pleasant :)

I also agree on the `rethinkdb.exe --install ServiceName --myFlags`, this looks great.

To give more broad ideas, you can also take a look at how the [other NoSQL...](https://docs.mongodb.com/manual/tutorial/install-mongodb-on-windows/#configure-a-windows-service-for-mongodb-community-edition) have implemented Windows Services. When I tried them, it was super easy to setup and their doc explains it very well. Their way of having an optional config file (instead of retyping the `--allFlags` every time) is also a very nice feature. I personally prefer the config file instead of the flags, since it becomes super easy to copy the exact same config to another Windows machine afterward.

Finally, you should probably use `--remove` instead of `--uninstall` since the word uninstall is generally use for uninstalling an App on Windows. So it might look like this instead:
```bash
rethinkdb.exe --install --myflags 
rethinkdb.exe --remove --myflags 
```

**Comment by danielmewes** [comment]
When installing as a service, we should consider requiring the use of a configuration file rather than allowing arbitrary parameters to be passed to `--install` directly. This simplifies the argument parsing on the server and might be clearer.

The syntax I imagine would be something like this:
```
rethinkdb.exe install-service ServiceName ConfigFilePath
rethinkdb.exe remove-service ServiceName
```

What do you think @bchavez @ghiscoding ?

**Comment by bchavez** [comment]
I like the idea of a configuration file. Makes lots of sense. :+1: The most near and dear to my heart is a drama-free `.zip+.exe` file or some (`installer.exe /extract`) option to extract the `rethinkdb.exe` executable.

*@bchavez just feels jaded when it comes to installers that integrate with M$ Windowz.* :scream_cat:

:running: :sunrise_over_mountains: [***And I ran. I ran so far away. I just ran. I ran all night and day. I couldn't get away...***](https://www.youtube.com/watch?v=0_Pq0xYr3L4)

**Comment by ghiscoding** [comment]
@danielmewes your last suggestions looks good, though I might change them a little to give more options to the user. I see that you already have a [.conf file](https://rethinkdb.com/docs/config-file/) available, so the concept would be to simply add extra settings for the Service inside the same file, no need to create 2 files that basically do similar things. 

**Option 1, with Config file**, also note that the config file should have higher priority over any other flags that might be passed through the command line. 
```bash
rethinkdb.exe --install-service ServiceName --config-file ConfigFilePath
rethinkdb.exe --remove-service ServiceName
```

**Option 2, without Config file**, that would be 0+ flags passed in
```bash
rethinkdb.exe --install-service ServiceName --flag1 --flag2
rethinkdb.exe --remove-service ServiceName
```

It's more flexible to have both options working. I also see that you already have a [.conf file](https://rethinkdb.com/docs/config-file/) available, so the concept would be to simply add these extra Service settings inside the same config file, no need to create 2 files that basically do similar things.

**Comment by danielmewes** [comment]
Windows service compatibility with `install-service` and `remove-service` commands is now implemented in the branch daniel_windows_service and in internal code review 3674.

A configuration file is required, to simplify argument parsing (otherwise we would have to distinguish arguments that apply to the `install-service` command itself and arguments that apply to the rethinkdb server, which would be more annoying).

**Comment by danielmewes** [comment]
(and the configuration file is the same we already have. There is no special configuration file for `install-service` itself. The file passed to `install-service` is simply passed on to the server when it gets run as the service.)

**Comment by danielmewes** [comment]
Merged into `next` via commit 6455c90 . This will ship with RethinkDB 2.4.0.

Some information on usage: https://github.com/rethinkdb/docs/issues/1146

**Comment by judwhite** [comment]
Will `rethinkdb.exe --install --myflags` be required to run as a Windows service, or will it be possible to just `sc create` the service?

**Comment by VeXocide** [comment]
It looks like `install_windows_service` in https://github.com/rethinkdb/rethinkdb/blob/e9658bb687d4ef8f546db5ced082938d91174956/src/clustering/administration/main/windows_service.cc#L127 does very little more than just install the service, thus it should be possible to do so via `sc create`. Make sure it has the right name though if you want to be able to remove it via `rethinkdb.exe`.

**Comment by danielmewes** [comment]
@judwhite Note that you'll have to pass special arguments to `rethinkdb.exe` in the service though to make it behave as a Windows service. Specifically it has to be run as `rethinkdb.exe run-service <server arguments>`. I'd definitely recommend using the `rethinkdb.exe install-service` option, unless you have a strong reason for using `sc create` instead.

**Comment by judwhite** [comment]
Thanks, @danielmewes. Typically we have other `sc` work like `sc config` to set the user identity and `sc failure` to setup actions in the case of a process crash. The compelling reason, maybe, is we wouldn't want the service to start until we've configured the identity under which it'll run.

I'm guessing `run-service` causes rethinkdb to handle stop/shutdown/interrogate events. If you want the process to detect if it's running as a service without the command line flag, you can query this with [GetTokenInformation](http://stackoverflow.com/questions/2668851/how-do-i-detect-that-my-application-is-running-as-service-or-in-an-interactive-s). I apologise if you've already considered this. Thanks again.

**Comment by Spiralis** [comment]
Was this shipped in 2.3.6, or is it still something pending for 2.4?

**Comment by ghiscoding** [comment]
@Spiralis, last I heard it will be part of `2.4` and isn't part of `2.3.6`, so we gotta wait patiently.
